File: src/symmetry.cpp
#include "symmetry.h"

#include <spglib.h>

#include <array>
#include <iomanip>
#include <iostream>
#include <map>
#include <memory>
#include <optional>
#include <set>
#include <vector>

#include "poscar_file.h"

void initializeSpglibInput(POSCAR& poscarDirect, double lattice[3][3], std::vector<std::array<double, 3>>& positions,
                           std::vector<int>& types, std::map<std::string, int>& element_map) {
    for (int i = 0; i < 3; ++i)
        for (int j = 0; j < 3; ++j)
            lattice[i][j] = poscarDirect.lattice[i][j];

    // Prepare atomic positions as flat vector (num_atoms*3)
    for (int i = 0; i < poscarDirect.total_atoms; ++i) {
        positions[i][0] = poscarDirect.coordinates[i].x;
        positions[i][1] = poscarDirect.coordinates[i].y;
        positions[i][2] = poscarDirect.coordinates[i].z;
    }

    // Prepare atomic types
    int type_counter{1};
    int idx{0};

    for (size_t i = 0; i < poscarDirect.elements.size(); ++i) {
        const std::string& el = poscarDirect.elements[i];
        int n = poscarDirect.num_atoms[i];
        if (element_map.find(el) == element_map.end())
            element_map[el] = type_counter++;
        for (int j = 0; j < n; ++j)
            types[idx++] = element_map[el];
    }
}

SpglibDatasetPtr analyzeSymmetry(const POSCAR& poscar, const double& symprec) {
    // Make a copy in fractional coordinates
    POSCAR poscarDirect = poscar;
    if (!poscarDirect.is_direct) {
        poscarDirect.toDirect();
    }

    int num_atoms = static_cast<int>(poscarDirect.total_atoms);
    double lattice[3][3];
    std::vector<std::array<double, 3>> positions(num_atoms);
    std::vector<int> types(num_atoms);
    std::map<std::string, int> element_map;

    initializeSpglibInput(poscarDirect, lattice, positions, types, element_map);

    SpglibDataset* dataset = spg_get_dataset(lattice, (double(*)[3])positions.data(), types.data(), num_atoms, symprec);

    return SpglibDatasetPtr(dataset, &spg_free_dataset);
}

std::optional<POSCAR> standardizeCell(const POSCAR& poscar, const double& symprec, const int& primitive,
                                      const int& idealize) {
    // 1) Copy and ensure fractional coordinates
    POSCAR poscarDirect = poscar;
    if (!poscarDirect.is_direct) {
        poscarDirect.toDirect();
    }

    int num_atoms_in = static_cast<int>(poscarDirect.total_atoms);
    int num_atoms{num_atoms_in * 64};

    double lattice[3][3];
    std::vector<std::array<double, 3>> positions(num_atoms);
    std::vector<int> types(num_atoms);
    std::map<std::string, int> element_map;

    initializeSpglibInput(poscarDirect, lattice, positions, types, element_map);

    // Checking if empty spheres are present in input
    for (const auto& el : poscar.elements) {
        if (el == "X" || el == "E" || el == "V" || el == "Vac") {
            std::cerr << "Warning: Empty sphere detected.\n"
                      << "SPGLIB will treat them as real atoms and symmetry may change.\n";
        }
    }

    // 3) Call spglib standardization
    int num_std = spg_standardize_cell(lattice, (double(*)[3])positions.data(), types.data(), num_atoms_in, primitive,
                                       idealize, symprec);

    if (num_std <= 0) {
        return std::nullopt;
    }

    // Determine maximum type index
    int max_type = 0;
    for (int i = 0; i < num_std; ++i)
        if (types[i] > max_type)
            max_type = types[i];

    // Create lookup vector (type -> element)
    std::vector<std::string> type_to_element(max_type + 1);

    // Fill from original element_map
    for (const auto& [el, t] : element_map) {
        if (t <= max_type)
            type_to_element[t] = el;
    }

    // 4) Build POSCAR from arrays now holding primitive cell
    POSCAR std_poscar;
    std_poscar.is_direct = true;
    std_poscar.total_atoms = num_std;
    std_poscar.comment = poscar.comment + (primitive == 1 ? " primitive cell" : " conventional cell");

    // Copy new lattice
    for (int i = 0; i < 3; ++i)
        for (int j = 0; j < 3; ++j)
            std_poscar.lattice[i][j] = lattice[i][j];

    // Copy positions
    std_poscar.elements.clear();
    std_poscar.num_atoms.clear();
    std_poscar.coordinates.clear();

    // Adding atomic coordinates to kepp the elemet order as well as correct assignment of atomic position to each
    // element
    for (const auto& original_el : poscar.elements) {
        int count_for_this_el = 0;

        for (int i = 0; i < num_std; ++i) {
            if (type_to_element[types[i]] == original_el) {
                Atom a;
                a.x = positions[i][0];
                a.y = positions[i][1];
                a.z = positions[i][2];
                std_poscar.coordinates.push_back(a);
                count_for_this_el++;
            }
        }

        // Pokud prvek v nové struktuře existuje, přidáme ho do hlavičky
        if (count_for_this_el > 0) {
            std_poscar.elements.push_back(original_el);
            std_poscar.num_atoms.push_back(count_for_this_el);
        }
    }
    return std_poscar;
}

void printSymmetryInfo(const SpglibDataset& dataset, const bool& wyckoff, const bool& symoperation) {
    std::cout << "=== Symmetry Information ===\n";

    // Space group
    std::cout << "Space group number: " << dataset.spacegroup_number << "\n";
    std::cout << "International symbol: " << dataset.international_symbol << "\n";

    // Hall symbol
    std::cout << "Hall symbol: " << dataset.hall_symbol << "\n";

    // Point group
    std::cout << "Point group: " << dataset.pointgroup_symbol << "\n";

    // Symmetry operations
    std::cout << "\nNumber of symmetry operations: " << dataset.n_operations << "\n";

    if (symoperation)
        printSymmetryOperations(dataset);

    // Compute number of irreducible atoms
    std::set<int> irreducible_atoms;
    for (int i = 0; i < dataset.n_atoms; ++i)
        irreducible_atoms.insert(dataset.equivalent_atoms[i]);

    std::cout << "\nNumber of Wyckoff positions (irreducible atoms): " << irreducible_atoms.size() << "\n";

    // Print Wyckoff letters for each atom
    if (wyckoff) {
        std::cout << "Wyckoff letters: ";
        for (int i = 0; i < dataset.n_atoms; ++i) {
            char wyckoff_letter = 'a' + dataset.wyckoffs[i];  // convert 0->'a', 1->'b', ...
            std::cout << wyckoff_letter << " ";
        }
    }
    std::cout << "\n";
}

void printSymmetryOperations(const SpglibDataset& dataset) {
    for (int i = 0; i < dataset.n_operations; ++i) {
        std::cout << "Operation " << i + 1 << ":\n";
        std::cout << "  Rotation matrix:\n";
        for (int j = 0; j < 3; ++j)
            std::cout << "   " << std::setw(2) << dataset.rotations[i][j][0] << " " << std::setw(2)
                      << dataset.rotations[i][j][1] << " " << std::setw(2) << dataset.rotations[i][j][2] << "\n";
        std::cout << "  Translation vector: " << dataset.translations[i][0] << " " << dataset.translations[i][1] << " "
                  << dataset.translations[i][2] << "\n";
    }
}

File: rust/src/symmetry.rs
//! Symmetry analysis and cell standardisation via the [`spglib`] Rust crate.
//!
//! [`spglib::dataset::Dataset`] is re-exported as the primary result type so
//! callers do not need a direct dependency on the `spglib` crate.

use std::collections::{BTreeMap, BTreeSet};
use std::os::raw::c_int;

use spglib::cell::Cell;
pub use spglib::dataset::Dataset;
use spglib_sys as ffi;

use crate::poscar::{Atom, Poscar};

// ─── Internal helpers ─────────────────────────────────────────────────────────

/// Build the spglib `Cell` from a POSCAR that is already in fractional
/// coordinates.  Also returns the element → integer-type mapping so callers
/// can reconstruct element names from type ids.
///
/// spglib uses **column** vectors (`lat[i][j]` = i-th Cartesian component of
/// the j-th basis vector), whereas POSCAR uses **row** vectors, so we
/// transpose here.
fn poscar_to_cell(poscar: &Poscar) -> (Cell, BTreeMap<String, i32>) {
    let mut lat = [[0.0f64; 3]; 3];
    for i in 0..3 {
        for j in 0..3 {
            lat[i][j] = poscar.lattice[j][i];
        }
    }

    let positions: Vec<[f64; 3]> = poscar
        .coordinates
        .iter()
        .map(|a| [a.x, a.y, a.z])
        .collect();

    let (types, element_map) = build_types(poscar);
    (Cell::new(&lat, &positions, &types), element_map)
}

/// Assign a unique integer type id to each distinct element symbol.
fn build_types(poscar: &Poscar) -> (Vec<i32>, BTreeMap<String, i32>) {
    let mut element_map: BTreeMap<String, i32> = BTreeMap::new();
    let mut counter: i32 = 1;
    let mut types = Vec::with_capacity(poscar.total_atoms as usize);

    for (el_raw, &count) in poscar.elements.iter().zip(poscar.num_atoms.iter()) {
        // Strip whitespace and slashes (mirrors the C++ implementation)
        let el: String = el_raw
            .chars()
            .filter(|&c| c != '/' && !c.is_whitespace())
            .collect();

        let type_id = *element_map.entry(el).or_insert_with(|| {
            let id = counter;
            counter += 1;
            id
        });
        for _ in 0..count {
            types.push(type_id);
        }
    }

    (types, element_map)
}

// ─── Public API ───────────────────────────────────────────────────────────────

/// Run spglib symmetry analysis and return a [`Dataset`] with all fields
/// decoded into owned Rust types.
///
/// Returns `None` if spglib returns a null dataset (symmetry search failed).
pub fn analyze_symmetry(poscar: &Poscar, symprec: f64) -> Option<Dataset> {
    let mut base = poscar.clone();
    if !base.is_direct {
        base.to_direct();
    }

    let (cell, _) = poscar_to_cell(&base);

    // Call spg_get_dataset through spglib_sys so we can check for null before
    // handing the pointer to Dataset::try_from (which would panic on null).
    let raw = unsafe {
        ffi::spg_get_dataset(
            cell.lattice.as_ptr() as *mut [f64; 3],
            cell.positions.as_ptr() as *mut [f64; 3],
            cell.types.as_ptr() as *const c_int,
            cell.positions.len() as c_int,
            symprec,
        )
    };

    if raw.is_null() {
        return None;
    }

    // Dataset::try_from transfers ownership of spglib's inner arrays into Vecs.
    Dataset::try_from(raw).ok()
}

/// Standardise a crystal cell using spglib.
///
/// * `primitive = true`  → primitive cell
/// * `primitive = false` → conventional cell
/// * `idealize`          → whether to idealise lattice parameters
///
/// Returns `None` if spglib fails.
///
/// # Note on memory
/// `spglib::cell::Cell::standardize` does not expose the new atom count and
/// does not pre-allocate for larger conventional cells, so we call
/// `spglib_sys::spg_standardize_cell` directly here with a properly-sized
/// buffer.
pub fn standardize_cell(
    poscar: &Poscar,
    symprec: f64,
    primitive: bool,
    idealize: bool,
) -> Option<Poscar> {
    // Warn about empty spheres
    for el in &poscar.elements {
        if matches!(el.as_str(), "X" | "E" | "V" | "Vac") {
            eprintln!(
                "Warning: empty sphere '{el}' detected; \
                 spglib will treat it as a real atom."
            );
        }
    }

    let mut base = poscar.clone();
    if !base.is_direct {
        base.to_direct();
    }

    let (cell_base, element_map) = poscar_to_cell(&base);
    let n_in = base.total_atoms as usize;

    // Pre-allocate 64× the input for worst-case conventional cell expansion.
    let capacity = n_in * 64;
    let mut lat = cell_base.lattice;
    let mut positions: Vec<[f64; 3]> = cell_base.positions.clone();
    positions.resize(capacity, [0.0; 3]);
    let mut types: Vec<i32> = cell_base.types.clone();
    types.resize(capacity, 0);

    let n_out = unsafe {
        ffi::spg_standardize_cell(
            lat.as_mut_ptr(),
            positions.as_mut_ptr(),
            types.as_mut_ptr() as *mut c_int,
            n_in as c_int,
            if primitive { 1 } else { 0 },
            if idealize { 1 } else { 0 },
            symprec,
        )
    };

    if n_out <= 0 {
        return None;
    }
    let n_out = n_out as usize;

    // Build reverse map: type_id → element symbol
    let max_type = types[..n_out].iter().copied().max().unwrap_or(0) as usize;
    let mut type_to_el = vec![String::new(); max_type + 1];
    for (el, &t) in &element_map {
        if (t as usize) <= max_type {
            type_to_el[t as usize] = el.clone();
        }
    }

    // Transpose spglib column-vector lattice back to POSCAR row-vector layout
    let mut out_lat = [[0.0f64; 3]; 3];
    for i in 0..3 {
        for j in 0..3 {
            out_lat[i][j] = lat[j][i];
        }
    }

    let suffix = if primitive { " primitive cell" } else { " conventional cell" };

    // Build output POSCAR, preserving the original element order
    let mut out_elements: Vec<String> = Vec::new();
    let mut out_num_atoms: Vec<i32> = Vec::new();
    let mut out_coords: Vec<Atom> = Vec::new();

    for orig_el in &poscar.elements {
        let mut count = 0i32;
        for i in 0..n_out {
            if type_to_el[types[i] as usize] == *orig_el {
                out_coords.push(Atom::new(positions[i][0], positions[i][1], positions[i][2]));
                count += 1;
            }
        }
        if count > 0 {
            out_elements.push(orig_el.clone());
            out_num_atoms.push(count);
        }
    }

    Some(Poscar {
        comment: format!("{}{suffix}", poscar.comment),
        scale: 1.0,
        lattice: out_lat,
        elements: out_elements,
        num_atoms: out_num_atoms,
        selective_dynamics: false,
        is_direct: true,
        total_atoms: n_out as i32,
        coordinates: out_coords,
    })
}

// ─── Printing ─────────────────────────────────────────────────────────────────

/// Print a human-readable symmetry summary to stdout.
pub fn print_symmetry_info(dataset: &Dataset, wyckoff: bool, symoperations: bool) {
    println!("=== Symmetry Information ===");
    println!("Space group number: {}", dataset.spacegroup_number);
    println!("International symbol: {}", dataset.international_symbol);
    println!("Hall symbol: {}", dataset.hall_symbol);
    println!("Point group: {}", dataset.pointgroup_symbol);
    println!("\nNumber of symmetry operations: {}", dataset.n_operations);

    if symoperations {
        print_symmetry_operations(dataset);
    }

    let irreducible: BTreeSet<i32> = dataset.equivalent_atoms.iter().copied().collect();
    println!(
        "\nNumber of Wyckoff positions (irreducible atoms): {}",
        irreducible.len()
    );

    if wyckoff {
        print!("Wyckoff letters: ");
        for &w in &dataset.wyckoffs {
            print!("{} ", (b'a' + w as u8) as char);
        }
        println!();
    }
}

/// Print all symmetry operations (rotation matrices + translation vectors).
pub fn print_symmetry_operations(dataset: &Dataset) {
    for (i, (rot, tr)) in dataset
        .rotations
        .iter()
        .zip(dataset.translations.iter())
        .enumerate()
    {
        println!("Operation {}:", i + 1);
        println!("  Rotation matrix:");
        for row in rot {
            println!("   {:2} {:2} {:2}", row[0], row[1], row[2]);
        }
        println!("  Translation vector: {} {} {}", tr[0], tr[1], tr[2]);
    }
}
